data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/SqsDynamicVisibilityTimeoutManager.java

```diff
@@ -0,0 +1,123 @@
+    }
+
+
+
+    public void startDynamicVisibilityForMessage(final Message message, final long objectSizeBytes) {
+        final int visibilityTimeoutIncrease = calculateVisibilityTimeoutFromObjectSize(objectSizeBytes);
+        final int waitBeforeIncreasingAgain = Math.min(MINIMUM_WAIT_BEFORE_CHANGE_VISIBILITY_TIMEOUT_SECONDS,
+                visibilityTimeoutIncrease - TIME_BETWEEN_CHANGE_MESSAGE_CALL_AND_VISIBILITY_TIMEOUT_SECONDS);
+
+        final ScheduledFuture<?> scheduledFuture = scheduledExecutorService.scheduleAtFixedRate(
+                () -> increaseVisibilityTimeout(visibilityTimeoutIncrease, message),
+                initialWaitBeforeIncrease,
+                waitBeforeIncreasingAgain, TimeUnit.SECONDS);
+
+        messageToScheduledFuture.put(message.messageId(), scheduledFuture);
+    }
+
+    public void stopDynamicVisibilityForMessage(final String messageId) {
+        final ScheduledFuture<?> scheduledFutureToCancel = messageToScheduledFuture.get(messageId);
+        if (Objects.nonNull(scheduledFutureToCancel)) {
+            boolean cancelled = scheduledFutureToCancel.cancel(true);
+
+            if (!cancelled) {
+                LOG.error("Unable to cancel dynamic visibility execution for message {}", messageId);
+            } else {
+                messageToScheduledFuture.remove(messageId);
+            }
+        }
+    }
+
+    public int getRemainingTimeBeforeNextChangeMessageIncrease(final String messageId) {
+        final ScheduledFuture<?> scheduledFutureToCancel = messageToScheduledFuture.get(messageId);
+        if (Objects.nonNull(scheduledFutureToCancel)) {
+            return (int) scheduledFutureToCancel.getDelay(TimeUnit.SECONDS);
+        }
+
+        return initialMessageVisibilityTimeout;
+    }
+
+    public void shutdownDynamicMessaging() {
+        try {
+            scheduledExecutorService.shutdown();
+            scheduledExecutorService.awaitTermination(3, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            LOG.error("Dynamic SQS messaging was interrupted during shutdown");
+        } finally {
+            if (!scheduledExecutorService.isTerminated()) {
+                scheduledExecutorService.shutdown();
+            }
+        }
+    }
+
+    private void increaseVisibilityTimeout(final int visibilityTimeoutIncrease,
+                                           final Message message) {
+
+        final ChangeMessageVisibilityRequest changeMessageVisibilityRequest = ChangeMessageVisibilityRequest.builder()
+                .receiptHandle(message.receiptHandle())
+                .visibilityTimeout(visibilityTimeoutIncrease)
+                .queueUrl(queueUrl)
+                .build();
+
+        try {
+            sqsClient.changeMessageVisibility(changeMessageVisibilityRequest);
+        } catch (final Exception e) {
+            LOG.error("An exception occurred while changing the visibility timeout for message {}", message.messageId(), e);
```

Comment:
We might need to kill the executor after X number of retries or some decision based logic. If a user purges a queue for example, the messages are all gone and we would presumably catch an exception and log here until data prepper shuts down, essentially a rogue thread

Reply:
If this happened, wouldn't the `SqsWorker` most likely be the first to know, and it could make a `stopDynamic` call

Reply:
The behavior I've seen is that the object worker class throws a NoSuchKeyException. If that is being bubbled up to the SqsWorker and handled by the existing logic then we should be good.